archive/phar_without_stub.ksy

```diff
@@ -0,0 +1,217 @@
+  
+  The phar stub usually contains code that loads the phar and runs a contained PHP file, but this is not required. A minimal valid phar stub is `<?php __HALT_COMPILER();` - such a stub makes it impossible to execute the phar directly, but still allows loading or manipulating it using the phar extension.
+doc-ref:
+  - 'https://www.php.net/manual/en/phar.fileformat.php'
+  - 'https://github.com/php/php-src/tree/master/ext/phar'
+  - 'https://svn.php.net/viewvc/pecl/phar/'
+  - 'https://svn.php.net/viewvc/pear/packages/PHP_Archive/'
+seq:
+  - id: manifest
+    type: manifest
+    doc: The archive's manifest, containing general metadata about the archive and its files.
+  - id: files
+    size: manifest.file_entries[_index].compressed_size
+    repeat: expr
+    repeat-expr: manifest.file_count
+    doc: The contents of each file in the archive (possibly compressed, as indicated by the file's flags in the manifest). The files are stored in the same order as they appear in the manifest.
+  - id: signature
+    type: signature
+    size-eos: true
+    if: manifest.flags.has_signature
+    doc: |
+      The archive's signature - a digest of all archive data before the signature itself.
+      
+      Note: Almost all of the available "signature" types are actually hashes, not signatures, and cannot be used to verify that the archive has not been tampered with. Only the OpenSSL signature type is a true cryptographic signature.
+enums:
+  signature_type:
+    0x1:
+      id: md5
+      -orig-id: PHAR_SIG_MD5
+      doc: Indicates an MD5 hash.
+    0x2:
+      id: sha1
+      -orig-id: PHAR_SIG_SHA1
+      doc: Indicates a SHA-1 hash.
+    0x4:
+      id: sha256
+      -orig-id: PHAR_SIG_SHA256
+      doc: Indicates a SHA-256 hash. Available since API version 1.1.0, PHP_Archive 0.12.0 and phar extension 1.1.0.
+    0x8:
+      id: sha512
+      -orig-id: PHAR_SIG_SHA512
+      doc: Indicates a SHA-512 hash. Available since API version 1.1.0, PHP_Archive 0.12.0 and phar extension 1.1.0.
+    0x10:
+      id: openssl
+      -orig-id: PHAR_SIG_OPENSSL
+      doc: |
+        Indicates an OpenSSL signature. Available since API version 1.1.1. This type is not documented in the phar extension's documentation of the phar format.
+        
+        Note: In older versions of the phar extension, this value was used for an undocumented and unimplemented "PGP" signature type (`PHAR_SIG_PGP`).
+types:
+  serialized_value:
+    seq:
+      - id: raw
+        size-eos: true
+        doc: The serialized value, as a raw byte array.
+    instances:
+      parsed:
+        pos: 0
+        type: php_serialized_value
+        doc: The serialized value, parsed as a structure.
+  file_flags:
+    seq:
+      - id: value
+        type: u4
+        doc: The unparsed flag bits.
+    instances:
+      permissions:
```

Comment:
Why not a bit field type?

Reply:
The official documentation specifies them as integer bit masks (see [here](https://www.php.net/manual/en/phar.fileformat.flags.php) and [here](https://www.php.net/manual/en/phar.fileformat.manifestfile.php)). Translating hex masks to bit fields is already not obvious because of bit order issues. In addition, here the flags are stored as a *little-endian* integer, so you also have to deal with the bytes being swapped (compared to the hex masks, which are naturally written in big-endian order).

The `permissions` field is a particular problem, because it spans two bytes. In combination with the little-endian byte swapping, this means that the `permissions` field would need to be split into two bit fields (the 8 least significant bits, then 7 unused bits, then the 1 most significant bit), so you would need to reconstruct it using a value instance anyway. kaitai-io/kaitai_struct#155 is probably relevant here.

This explanation might not be completely accurate, I might be misremembering some things. A while ago I tried implementing this using bit fields and couldn't get it to work after many tries, so I changed it to the integer and bit mask combination and it worked on the first try. Since the official format documentation and the two major implementations treat these fields as integers with bit masks, I think it makes sense for our spec to do the same, especially since the alternative with bit fields would be a pain to write and hard to understand.

Reply:
I don't think using bit fields here makes sense until Kaitai Struct allows switching to LSB-first bit order. Processing little-endian data in MSB-first bit order is too confusing IMHO, and as far as I can tell doesn't offer any big advantage over the current integer mask approach.